keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,744 @@
+
+Unfortunately, the current behavior is inconsistent with how we treat terminating and terminated pods
+in the deployment controller.
+
+- The Recreate Deployment strategy waits for terminating pods to terminate before creating
+  (scheduling) new pods.
+- The RollingUpdate deployment strategy does not wait for terminating pods and creates (schedules)
+  new pods immediately.
+- Scaling up a Deployment also does not wait for terminating pods and creates (schedules) new pods
+  right away.
+
+Unfortunately, in Deployments with a Recreate strategy we can get mixed behavior. The
+deployment will wait for old pods to terminate during a rollout, but will ignore the terminating
+pods when scaling the pods. So it is still possible to end up with a larger number of pods than
+`.spec.replicas`.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an application user, I would prefer predictable number of pods in my cluster to prevent any
+scheduling issues and unnecessary autoscaling of nodes. I would also like to achieve consistent
+allocation of other scarce resources to pods.
+
+#### Story 2
+
+As an application user, I would like to keep the old behavior of fast scaling of pods and do not
+mind the higher utilization of resources.
+
+#### Story 3
+
+As an application user, I would like to track the number of instances that perform useful work
+during the entire lifecycle of a pod.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Consideration for Other Controllers
+
+This feature is not considered for standalone ReplicaSets except for tracking the terminating pods.
+The reason for this is that ReplicaSet behavior is meant to be simple and used as a building block
+by other high-level controllers. If we included the PodReplacementPolicy in both ReplicaSets and
+Deployments, it would be hard to reconcile these fields because a ReplicaSet only has the local view
+of its own pods. The Deployment has the complete picture of all the pods (through ReplicaSet's status)
+in its ReplicaSets and can make the correct balancing decision. Adding such a feature to ReplicaSets
+could also pose a threat to third-party controllers that embed ReplicaSets in their resource
+definitions, as this could alter their behavior.
+
+This feature is also not desirable for StatefulSets and DaemonSets, because by design we wait until
+old pods terminate before creating new pods.
+
+This feature is already implemented for Jobs.
+
+### Risks and Mitigations
+
+Deployment rollouts might be slower when using the `Terminated` PodReplacementPolicy.
+
+Deployment rollouts might consume excessive resources when using the `Terminating` PodReplacementPolicy.
+
+This is mitigated by making this feature opt-in.
+
+## Design Details
+
+### Deployment Behavior Changes
+
+Recreate rollout logic:
+- Terminating:
+    1. Scale down old ReplicaSet(s) to 0.
+    2. Wait until all the pods are at least terminating.
+    3. Create new replica set.
+- Terminated: Current behaviour.
```

Comment:
I am assuming the plan is also to make this the default behaviour, for obvious reasons. However, I think it makes sense to explicitly say it here.
```suggestion
- Terminated: Current behaviour. `Terminated` will be the default for PodReplacementPolicy when using Recreate rollout strategy.
```

Reply:
I think we should not / cannot default the value according to the rollout strategy.

- The field should be independent from the rollout strategy as much as possible . So this means we should not change the field when the rollout strategy changes. If we set this on create admission according to the strategy, we might make the behaviour backwards incompatible. That is if someone's flow involves changing the strategy later after the creation.
- the scaling behaviour is not defined anywhere and if we set the PodReplacementPolicy to `Terminated` when observing the Recreate rollout strategy, we would start applying that to the scaling behaviour as well. This would also be backwards incompatible.

To sum up, we need an option that preserves today's mixed behaviour.

Reply:
That makes more sense indeed. Should `basically it means we do not touch the current behaviour when the field is nil` go into the KEP somewhere then?

Reply:
It is mentioned in the API. Let me know if any part is not clear.

```
// The default behavior when the policy is not specified depends on the DeploymentStrategy:
// - Recreate strategy uses Terminated behavior when recreating the deployment,
//   but uses Terminating when scaling the deployment.
// - RollingUpdate strategy uses Terminating behavior for both rolling out and
//   scaling the deployments.
```

Reply:
That's true, looks good :rocket: